New GUI front-end

jpt13653903 opened the pull request:
Issue: #10 

A pre-emptive pull-request in order to obtain email-notifications on commits to this branch.

jpt13653903 committed:
Updated the how-to in the USB PIC example

Commit:
Moved illustrations to a new folder

Commit:
Gitignore

jpt13653903 commented:
Think you can work [the new feature](https://github.com/jpt13653903/Gerber2PDF/pull/14) into the new GUI?

Commit:
Unfinished GTK3 version.

Stuck at GObject.
Simple POD structs need to be wrapped in GObject boilerplate for
adding to Gio::ListStore :(

Commit:
Trying to overcome GObject errors. Still stuck

Commit:
ImGui skeleton

Commit:
Basic GUI skeleton using Dear ImGui for C++

Commit:
Merge remote to local

jpt13653903 commented:
First time using Meson -- I can see the appeal :-)

I'm having some trouble building with it...  How do I specify a search path for libraries?

It's complaining that it cannot find GLFW, even though [it is there](https://github.com/ocornut/imgui/tree/50c43d1c8e031c56177fd13aab076ed3e1a75c26/examples/libs/glfw).

That doesn't come with the library binaries, but those are [easy enough to get](http://www.glfw.org/download.html) and install to where the compiler can find it (if I can provide the build system with a path...).

happycoder97 commented:
I haven't built it on Windows. The GUI is not complete yet.
Hopefully will complete the GUI today and build for windows tomorrow. :)

jpt13653903 commented:
> 
> I haven't built it on Windows. The GUI is not complete yet.
> Hopefully will complete the GUI today and build for windows tomorrow. :)

No worries -- I just wanted to get my build-chain set up so long.  It'll be amazing progress if it's "complete" by the end of today ;-)

The main source is not Linux-compatible at the moment.  I think it's only the [JFile class](https://github.com/jpt13653903/Gerber2PDF/tree/feature/GUI/Source/Tools/File) that is still obsolete -- the rest should be cross-platform, but I'm not sure...  I want to replace them with [the Toolbox](https://github.com/jpt13653903/Toolbox) anyway, and will do so soon -- after I've updated it to be platform independent.

Anyway.  You can delete your [Illustrations](https://github.com/jpt13653903/Gerber2PDF/tree/feature/GUI/GUI/Illustrations) folder and do any updates to the concept directly in [Docs/GUI/Concept](https://github.com/jpt13653903/Gerber2PDF/tree/feature/GUI/Docs/GUI/Concept).

I have a read-me there with a couple ideas and questions...

Commit:
Properly implemented gerber files list box

jpt13653903 commented:
I've ported the base `Source` to Linux...  See #15

Commit:
Improved meson.build for windows

Commit:
Refactored and improved Gerber List and FileChooser

happycoder97 commented:
I made some changes to meson.build file. 
Now it should build under MSYS2 MinGW environment.
The compiler provided by tdm-gcc was released in 2015.
I had used some C++17 features in the code. That might prevent it from compiling under tdm-gcc.
Should I rewrite it for C++11?

happycoder97 commented:
[Build.zip](https://github.com/jpt13653903/Gerber2PDF/files/2611975/Build.zip)
Here's the executable with required dependencies.

Still to implement: Background layer color, `-page_size` option, output file selection, execution of final batch script.
Currently, generated batch script is printed to terminal when clicking save batch file button.

happycoder97 commented:
The combine checkbox is replaced with a new page separator. Thanks for the idea. It made the UI more intuitive.
The color button will let us input RGB and A also.
Background color selector is on the way.
For loading a saved configuration, let me think of some ways to re-parse the exported batch script.

jpt13653903 commented:
> I made some changes to meson.build file.
> Now it should build under MSYS2 MinGW environment.
> The compiler provided by tdm-gcc was released in 2015.
> I had used some C++17 features in the code. That might prevent it from compiling under tdm-gcc.
> Should I rewrite it for C++11?

Might not be a bad idea to down-port to C++14.  GCC version 5 (the current MinGW) [has full support](https://gcc.gnu.org/projects/cxx-status.html).

----------------------

> [Build.zip](https://github.com/jpt13653903/Gerber2PDF/files/2611975/Build.zip)
> Here's the executable with required dependencies.
>
> Still to implement: Background layer color, -page_size option, output file selection, execution of final batch script.
> Currently, generated batch script is printed to terminal when clicking save batch file button.

This is fine -- you are at prototype (pre-alpha) stage anyway.  There's lots of work left on the layout, but this can happen later (after all the functionality is in).

Feel free to merge `master` into your `feature/GUI` so that you can test the new stuff...  Also makes merging back to `master` easier later on.

----------------------

> The combine checkbox is replaced with a new page separator. Thanks for the idea. It made the UI more intuitive.
> The color button will let us input RGB and A also.
> Background color selector is on the way.
> For loading a saved configuration, let me think of some ways to re-parse the exported batch script.

Random idea for loading a saved configuration -- export a JSON string.  You can export to a separate file, or to a "rem" line in the batch script.

I wrote a little JSON abstraction some time ago -- submodule the [Toolbox](https://github.com/jpt13653903/Toolbox).  You can have a look at the [unit test](https://github.com/jpt13653903/Toolbox/blob/master/Test/testJSON.cpp) on how to use it.  It's not the most versatile or well-written library, so feel free to suggest a different one.

Commit:
Fixed:
Color buttons not working properly
Made file chooser address bar accept user input.
Aded background layer color, page size and stroke to fill options

jpt13653903 committed:
Added some ideas to the GUI concept design

jpt13653903 commented:
Some preliminary thoughts:

- I really like what you've done with it (conceptually).  I like that you're aiming for touch-screen compliance (big buttons, etc.).  Do try for consistency, however -- all buttons the same height, for instance.
- You can add [an icon](https://github.com/jpt13653903/Gerber2PDF/tree/master/Icon) (only in the window title bar -- it looks terrible when used as part of the GUI itself).
- You can remove the `*.gbr` filter.  Very few CAD tools export with the `gbr` extension.  Some export with `grb`, others with the layer name (`*.top`, `*.bot`, etc.).
- I would love to be able to drag-and-drop within the file list - that makes it easier to re-order the files.  In the interim (because getting drag-drop to work right is a bit of a pain) a simple "move up" and "move down" button will do.
- Try for cultural ambiguity (icons instead of words).  I'd like it to be multi-lingual without having to do any translation...  See [this idea](https://github.com/jpt13653903/Gerber2PDF/blob/feature/GUI/Docs/GUI/Concept/gui_draft_2.svg) for details (my icons are terrible, but you get the idea...).
- I REALLY like the colour pop-up info box :-)
- My converter will not like files starting with `file://`
- Last time I used GLFW it only supported one window...  I hope they've improved matters, because it would be really nice if the select-file dialogue box can float over the main window instead of inside it.
- Please add a "draw" trigger to the "on-resize" event -- very disconcerting when you resize the window and the content doesn't live-update.

jpt13653903 commented:
I'm starting to think it might be a good idea to re-factor my converter so that you can statically link the engine into the GUI.  This will make it easier for the user to set up the system (no need for special paths, etc.)

Eventually I might even want to link the two into a single project -- the GUI loads when you run the executable with no parameters; and the converter works as it does at the moment (i.e. terminal only) when you have parameters present.

Let me think on it a bit more...

jpt13653903 committed:
Merge branch 'master' into feature/GUI

happycoder97 commented:
>     * My converter will not like files starting with `file://`
Oops.. It was a dummy entry that I inserted for quick testing of UI.
The files loaded using the add files button will not have file:// prepended :)

happycoder97 commented:
>     * I REALLY like the colour pop-up info box :-)
Thanks to https://github.com/ocornut/imgui :-)

Commit:
Removed C++17 features, replaced them with Boost versions. Now compiles with C++14

Commit:
Merge branch 'feature/GUI' of github.com:jpt13653903/Gerber2PDF into feature/GUI

Commit:
Added linux to windows-i686 cross compilation options

Commit:
Added onwindowresize callback to redraw the window

happycoder97 commented:
[Build.zip](https://github.com/jpt13653903/Gerber2PDF/files/2612667/Build.zip)
Fixed resize issue, removed C++17 features.

Commit:
Fixed selecting save file.
Removed *.gbr filter
Added Letter option to Page Size

jpt13653903 commented:
Thanks for the build :-)  I haven't gotten round to setting up my build chain yet...

Small things I've noticed:

- Thanks for adding the draw-while-resizing trigger -- I always love how smooth OpenGL-based GUIs feel.
- When adding files, you default to mirrored...
- The background colour should default to transparent (#FFFFFF00)
- The page size should default to "tight" (this option is selected when the command-line option is absent)
- The background colour, page size and stroke to fills can be moved to the bottom of the GUI (or an "advanced" section / drawer).  These are options very few people actually use.
- When you add the other side-line options (silent exit, etc.), these can also go in the "advanced" section.
- When you get around to saving the shell / batch script...  Remember to add new-line escape characters.  Also: Windows batch has a stupid rule where the escape-new-line does not only escape the new-line, but also the first character on the next line, so remember to add indentation.  See [the example](https://github.com/jpt13653903/Gerber2PDF/tree/feature/GUI/Examples/USB%20PIC) for details.

jpt13653903 commented:
> I have been generating the id from file name.

Ah - that would explain why I was struggling to reproduce it...

jpt13653903 commented:
> I think I should use some index there.

I find that one of two options generally work for me...

- Use the pointer to the instance directly (generally a bad option if you plan on making the list dynamic, so it won't work here).
- Use a large global counter that increments every time you need a new ID (this works for small programs that don't run for very long, so this might work).
- Use a GUID from some or other generator library (the nuclear option).

Commit:
PageSize is now an enum class.
A helper function is defined in logic.hpp to convert it to string.
Minor UI Fixes.
Changed "Label##Id description" to "Label##ID_DESCRIPTION" in some widgets.

Commit:
Fixed the color button bug when same file is added multiple times.

Commit:
Improved insert page break handling.
Now appends page break at the end.
Fixed a bug which caused the batch script to insert \n^ instead of ^\n

Commit:
Added confirmation for save to file dialogs when file exists

Commit:
Fixed some comments

jpt13653903 commented:
My build chain is up and running :-)

I've given up on meson and wrote a Makefile instead (which I'll commit shortly).

- I've downloaded the newest TDM version of MinGW today (version `g++.exe (tdm64-1) 5.1.0`)
- I've also downloaded boost today (version 1.68.0) and built it using gcc, with C++14 enabled.
- I'm using a [pre-built](http://www.glfw.org/download.html) version of GLFW (version 3.2.1, 64-bit Windows).
- The OpenGL library is whatever comes with my screen card (nVidia GeForce GT 750M).

------------------------

Commit e179d85 builds and runs

------------------------

Commit d4ddebd builds, but exits with:

```
C:\JPT\Projects\12\1204 - Gerber2pdf\GUI\build.win>Gerber2pdf_GUI.exe
Assertion failed!

Program: C:\JPT\Projects\12\1204 - Gerber2pdf\GUI\build.win\Gerber2pdf_GUI.exe
File: imgui/imgui_draw.cpp, Line 1576

Expression: 0

C:\JPT\Projects\12\1204 - Gerber2pdf\GUI\build.win>
```

------------------------

All newer commits (up to and including 11082c9) fails to build:

```
rm -rf build
rm -rf build.win
mkdir -p build.win/obj
x86_64-w64-mingw32-g++ -Wall -fexceptions -O2 -std=c++14 -c -Iimgui -Iimgui/examples -Iimgui/misc/cpp -IC:/TDM-GCC-64/lib/glfw-3.2.1/include -IC:/TDM-GCC-64/lib/boost_1_68_0 main.cpp -o build.win/obj/main.o
mkdir -p build.win/obj
x86_64-w64-mingw32-g++ -Wall -fexceptions -O2 -std=c++14 -c -Iimgui -Iimgui/examples -Iimgui/misc/cpp -IC:/TDM-GCC-64/lib/glfw-3.2.1/include -IC:/TDM-GCC-64/lib/boost_1_68_0 gui_main.cpp -o build.win/obj/gui_main.o
gui_main.cpp: In function 'void render_main_action_btns(MainState*, ImVec2)':
gui_main.cpp:144:53: error: no matching function for call to 'std::basic_ofstream<char>::basic_ofstream(<brace-enclosed initializer list>)'
         std::ofstream ofstream{batch_file[0].c_str()};
                                                     ^
In file included from gui_main.cpp:6:0:
C:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/5.1.0/include/c++/fstream:725:7: note: candidate: std::basic_ofstream<_CharT, _Traits>::basic_ofstream(std::basic_ofstream<_CharT, _Traits>&&) [with _CharT = char; _Traits = std::char_traits<char>]
       basic_ofstream(basic_ofstream&& __rhs)
       ^
C:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/5.1.0/include/c++/fstream:725:7: note:   no known conversion for argument 1 from 'const value_type* {aka const wchar_t*}' to 'std::basic_ofstream<char>&&'
C:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/5.1.0/include/c++/fstream:715:7: note: candidate: std::basic_ofstream<_CharT, _Traits>::basic_ofstream(const string&, std::ios_base::openmode) [with _CharT = char; _Traits = std::char_traits<char>; std::__cxx11::string = std::__cxx11::basic_string<char>; std::ios_base::openmode = std::_Ios_Openmode]
       basic_ofstream(const std::string& __s,
       ^
C:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/5.1.0/include/c++/fstream:715:7: note:   no known conversion for argument 1 from 'const value_type* {aka const wchar_t*}' to 'const string& {aka const std::__cxx11::basic_string<char>&}'
C:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/5.1.0/include/c++/fstream:697:7: note: candidate: std::basic_ofstream<_CharT, _Traits>::basic_ofstream(const char*, std::ios_base::openmode) [with _CharT = char; _Traits = std::char_traits<char>; std::ios_base::openmode = std::_Ios_Openmode]
       basic_ofstream(const char* __s,
       ^
C:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/5.1.0/include/c++/fstream:697:7: note:   no known conversion for argument 1 from 'const value_type* {aka const wchar_t*}' to 'const char*'
C:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/5.1.0/include/c++/fstream:682:7: note: candidate: std::basic_ofstream<_CharT, _Traits>::basic_ofstream() [with _CharT = char; _Traits = std::char_traits<char>]
       basic_ofstream(): __ostream_type(), _M_filebuf()
       ^
C:/TDM-GCC-64/lib/gcc/x86_64-w64-mingw32/5.1.0/include/c++/fstream:682:7: note:   candidate expects 0 arguments, 1 provided
Makefile:67: recipe for target 'build.win/obj/gui_main.o' failed
make: *** [build.win/obj/gui_main.o] Error 1
```

-----------------

There are a couple of warnings relating to the use of `long` vs. `long long` and other integer types.  Keep in mind that on Windows-64, `long` is 32-bit, whereas on Linux-64 it's 64-bit.

This is especially problematic when trying the `printf` something of type `size_t`, like you're doing [here](https://github.com/jpt13653903/Gerber2PDF/blob/3762421e855ed57f528dcfebfdc9d71d886f79f5/GUI/gerber_list_box.cpp#L167).  Linux and Windows define the size_t type differently.  And it get's worse when you're also supporting 32-bit...

For debug printing this is fine, but it might be worth while to get into the habit of either explicit type-casting, or using the C++ libraries (eg. `cout  << "Delete##" << index`).

jpt13653903 committed:
Created a Makefile to build the GUI on Windows

jpt13653903 commented:
My apologies: commit d4ddebd does run -- you just need to have the font in the `fonts` folder, relative to the current directory.

It might be a safer option to extract the executable path from `argv[0]` and load resources relative to that...  Then the current directory doesn't matter.

happycoder97 commented:
`snprintf(..)` was a naive attempt to generate `id` for different entries.
I later figured out how to do it properly. Newer commits doesn't include printf's :)

Commit:
Fixed a bug due to windows' path being wchar_t

Commit:
Merge branch 'feature/GUI' of github.com:jpt13653903/Gerber2PDF into feature/GUI

jpt13653903 commented:
The latest commit compiles and runs without problems - thank you :-)

I'm busy with [the refactoring](https://github.com/jpt13653903/Gerber2PDF/tree/refactor/EngineToClass) -- soon you'll be able to link to `libEngine.a` and use the converter directly...

Commit:
gitignore visual studio workspace file

jpt13653903 committed:
Merge branch 'master' into feature/GUI

jpt13653903 commented:
Few small things on the output bat file...

- Single quotes in Windows is just another character -- use double quotes instead.
- The "output" option needs a "-" in front
- All Gerber files should be enclosed in double quotes (for in case the file names and / or paths have spaces)
- You have a forward-slash that sneaked in there (just after the drive specifier).

For now I think you should focus only on exporting Linux shell scripts (since you're testing in Linux).  That way you can run the script to find any snags with it.  Porting that to a bat file export is a simple case of swapping the slashes around and replacing the escape new-line characters.  You can do that with something like:

```C++
#ifdef __linux__
  #define PathSep '/'
  #define EscNL '\\'
  #define ScriptExt ".sh"
#else // Windows
  #define PathSep '\\'
  #define EscNL '^'
  #define ScriptExt ".bat"
#endif
```

jpt13653903 committed:
Merge branch 'master' into feature/GUI

jpt13653903 committed:
Added Linux scripts for the new examples

jpt13653903 committed:
Merge branch 'master' into feature/GUI

happycoder97 commented:
Sorry for the inactivity. I got stuck due to college stuff. I hope to resume the work next week.

jpt13653903 commented:
No worries - You're doing me a MASSIVE favour...

Time-lines are of no concern to me ;-)



On Fri, 30 Nov 2018, 13:36 Farzeen, <notifications@github.com> wrote:

> Sorry for the inactivity. I got stuck due to college stuff. I hope to
> resume the work next week.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jpt13653903/Gerber2PDF/pull/12#issuecomment-443177783>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AnUUd1Go0P93ooWJNViDiJHRS6jP3dhOks5u0RhBgaJpZM4YoCJg>
> .
>

Commit:
Merge branch 'feature/GUI' of github.com:jpt13653903/Gerber2PDF into feature/GUI

Commit:
File chooser: Clear selected rows after closing window

Commit:
Fixed: "-output" was missing a "-"
Fixed: Replace single quotes with double quotes

happycoder97 commented:
Hi, I'm back :)

I looked at Engine.h.
In the constructor, is `NumPages` equal to number of Gerber files, or can one Gerber file span multiple pages?

I should construct `Engine` once, then 
```
For each gerber file:
    Set the parameters like color, mirror, combine, new page etc 
    Call Run(filename, title)
````
then finally call `Finish(output_filename)`, right?

What is `title` in Run()? 
Shall I add a text field in the list entries to specify title?

Commit:
meson.build: added dependency to libEngine.a

jpt13653903 commented:
Welcome back :-)

You can look at how to use Engine.h in [main.cpp](https://github.com/jpt13653903/Gerber2PDF/blob/master/Source/main.cpp).  In summary:

In the constructor, you can set `NumPages` to any practical value.  This is typically larger than or equal to the number of output pages.  If you combine all the Gerber files onto a single page, set this to 1.  It simply does a pre-allocation of memory.  I set it to the number of arguments, because I'm too lazy to count the pages.  You can safely set it to the number of lines in your list.

One Gerber file can be put onto multiple pages (see [the USB PIC example](https://github.com/jpt13653903/Gerber2PDF/tree/master/Examples/USB%20PIC)).  It cannot "span" multiple pages.

You need to create a new instance of `ENGINE` every time you create a PDF.

The title is displayed in the PDF bookmark tree.  You can be fancy and specify the title in a text-box (that would be quite nice actually), but I just make it equal to the file name.  The `FileName` parameter might or might not include a path, which I don't want to put into the bookmark title.  The title is overwritten when the title is defined in the Gerber file (which is part of the newer standard).

Otherwise, yes &ndash; your algorithm looks right:

1. Create a new instance, which allocates memory for `NumPages` number of PDF pages
1. Set the public members as required, and call `Run` for each line on your list.
1. Call `Finish`
1. Destroy the instance.

Commit:
Implemented Execute button

Commit:
Fixed a bug, which ignored page breaks

Commit:
Added a flag to use mingw32 compiler for cross compilation from Linux to Windows

Commit:
Added a Readme explaining distribution of the program for Windows

jpt13653903 commented:
> `Engine.NewPage` and `Engine.Combine` are always opposite to each other, right?
> Is there any case they should be same?

`NewPage` takes priority, so they can be the same.  A new page is created when `NewPage` is true, or `Combine` is false.

Commit:
Fixed engine include paths for windows makefile

jpt13653903 commented:
One other thing you need to know about:  `NewPage` auto-clears (see [here](https://github.com/jpt13653903/Gerber2PDF/blob/3c0f92102856018764222f40fa67600156ecbb87/Source/Engine.cpp#L632))

The general idea is to keep `Combine` "true", and set `NewPage` whenever you have a page-break...

jpt13653903 committed:
Updated GUI/Makefile to link to libEngine

jpt13653903 commented:
> I think the project is at a stable enough place to warrant starting a Kanban board, so I've [set one up](https://github.com/jpt13653903/Gerber2PDF/projects/1) and added some cards.

I don't think we need to go all-out issues yet (it's too early), but normal cards might be useful in keeping track of early bugs and features...

jpt13653903 commented:
I've reorganised the [Kanban board](https://github.com/jpt13653903/Gerber2PDF/projects/1) a bit (split the low-priority cards into a separate column, added some more cards based on previous comments above, etc.).

Feel free to add cards of your own, drag the existing ones around, etc.

happycoder97 commented:
How did you install boost and glfw with tdm-gcc?
I am only used to building within cygwin and mingw-w64.
I tried downloading the boost-xxx.7z file from boost.org, extract and run bootstrap.bat.
It exited with error `cl.exe` not found error, as I don't have visual studio installed.
Same with GLFW
Isn't it possible to build both using tdm-gcc?

I have setup a windows vm. 
Building and running from windows will help me identify bugs faster.

happycoder97 commented:
PS: Running `bootstrap gcc` results in the following linker errors:
```
undefined reference to `UnregisterWait`
undefined reference to `RegisterWaitForSingleObject`
```

jpt13653903 commented:
It's a couple of special switches...

I'm out for the day - will come back to you this evening...

On Sat, 5 Jan 2019, 08:59 Farzeen, <notifications@github.com> wrote:

> PS: Running bootstrap gcc results in the following linker errors:
>
> undefined reference to `UnregisterWait`
> undefined reference to `RegisterWaitForSingleObject`
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jpt13653903/Gerber2PDF/pull/12#issuecomment-451633454>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AnUUd9OxVEji-ruVZojx_mZWT_r7oh_Aks5vAE08gaJpZM4YoCJg>
> .
>

Commit:
Accept incoming changes in Makefile

Commit:
Fix `C:/` forward slash in windows

Commit:
Saving Batch File:
Put gerber files in quotes
Remove newline from last line

Commit:
Disable mirroring by default

Commit:
Fixed: Extents page size option not showing

Commit:
Load fonts relative to the executable location, not relative to cwd.
Do not crash when fonts are not found. Log to the console instead.

Commit:
Save Batch File: If output pdf file is not specified, default to output.pdf.

Commit:
Switch from meson to Makefile for Linux

happycoder97 commented:
I have made some changes to GUI/Makefile for enabling compilation on both Windows and Linux.
I am not sure I did it correctly.. Could you please review?
If it is too messy we can revert it

jpt13653903 committed:
Merge branch 'master' into feature/GUI

jpt13653903 committed:
Fixed a few bugs in GUI/Makefile so that it builds in Windows

jpt13653903 commented:
I've fixed the Makefile to build on my system, but I'm getting a weird bug...

I've modified your `gui_setup` function to:

```C++
void gui_setup(int argc, char **argv) {
  printf("gui_setup\n"); fflush(stdout);
    ImGui::StyleColorsDark();
    ImGui::GetStyle().WindowRounding = 0.0f;

    auto io = ImGui::GetIO();
    auto prog_path = boost::dll::program_location().parent_path();
  wprintf(L"prog_path = %ls\n", prog_path.c_str()); fflush(stdout);
    auto font_path = prog_path / "fonts" / "FiraSans-Regular.ttf";
  wprintf(L"font_path = %ls\n", font_path.c_str()); fflush(stdout);
  printf("Here...1\n"); fflush(stdout);
    if(fs::exists(font_path)) {
  printf("Here...2\n"); fflush(stdout);
        auto font_path_str = fs::canonical(font_path).string();
  wprintf(L"font_path_str = %ls\n", font_path_str.c_str()); fflush(stdout);
  printf("Here...3\n"); fflush(stdout);
        g_imgui_view.normal_font = io.Fonts->AddFontFromFileTTF(font_path_str.c_str(), 14.0f);
  printf("Here...4\n"); fflush(stdout);
        g_imgui_view.title_font = io.Fonts->AddFontFromFileTTF(font_path_str.c_str(), 36.0f);
  printf("Here...5\n"); fflush(stdout);
    } else {
  printf("Here...6\n"); fflush(stdout);
        std::cout << "Warning: Couldn't load fonts. The UI may appear ugly. \n";
        std::cout << "   " << font_path << " not found\n";
    }
  printf("Here...7\n"); fflush(stdout);

    ui_state.active_row_index = -1;
}
```

When I build and run, it gives:

```
User@JPT-MSI MINGW64 /c/JPT/Projects/12/1204 - Gerber2pdf/GUI/dist/win64 (feature/GUI)
$ ./Gerber2pdf_GUI.exe
gui_setup
prog_path = C:\JPT\Projects\12\1204 - Gerber2pdf\GUI\dist\win64
font_path = C:\JPT\Projects\12\1204 - Gerber2pdf\GUI\dist\win64\fonts\FiraSans-Regular.ttf
Here...1
Here...2
terminate called after throwing an instance of 'boost::filesystem::filesystem_error'
  what():  boost::filesystem::read_symlink: The request is not supported: "C:/JPT\Projects\12\1204 - Gerber2pdf"

User@JPT-MSI MINGW64 /c/JPT/Projects/12/1204 - Gerber2pdf/GUI/dist/win64 (feature/GUI)
$
```

jpt13653903 commented:
> How did you install boost and glfw with tdm-gcc?
> I am only used to building within cygwin and mingw-w64.
> I tried downloading the boost-xxx.7z file from boost.org, extract and run bootstrap.bat.
> It exited with error `cl.exe` not found error, as I don't have visual studio installed.
> Same with GLFW
> Isn't it possible to build both using tdm-gcc?
> 
> I have setup a windows vm.
> Building and running from windows will help me identify bugs faster.

I haven't bothered trying the build `libglfw3`.  I simply use the pre-built binaries available from [the website](http://www.glfw.org/download.html).

To build Boost, I ran the following commands after extracting the boost source to `C:\TDM-GCC-64\lib\boost_1_68_0`:

```
bootstrap.bat gcc
b2 cxxstd=14
```

I don't have Visual Studio installed.  I use `git bash` as my terminal (I don't have CygWin, and don't like the native Windows console), and all compilation goes through MinGW.

jpt13653903 commented:
> terminate called after throwing an instance of 'boost::filesystem::filesystem_error' what(): boost::filesystem::read_symlink: The request is not supported: "C:/JPT\Projects\12\1204 - Gerber2pdf"

I have a sneaky suspicion...

My system is set up with symbolic links (or "directory junctions", as they are called in Windows).  the folder `1204 - Gerber2pdf` is set up to point to `W:\Projects\12\1204 - Gerber2pdf`.  It would seem as if the Boost `fs::canonical` function doesn't like that...

I've removed it, and now it works -- commit to follow shortly...

jpt13653903 committed:
Fixed bug relating to Windows directory junctions in the path to the font folder

jpt13653903 committed:
Merge branch 'master' into feature/GUI

jpt13653903 committed:
Updated the GUI code to be compatible with the new ENGINE constructor

jpt13653903 commented:
I fixed a bug in the engine and updated your code to be compatible.  Please pull the newest changes as soon as you're able...

jpt13653903 commented on GUI/Makefile:
```diff
@@ -164,6 +163,9 @@ $(DistDir)/$(EXEName): $(Objects) $(DLLs) $(Fonts)
 	$(CXX) $(Options) -c $(Includes) $< -o $@
+
+$(Engine)/obj/libEngine.a: FORCE
+	make -C $(Engine)
```
@happycoder97 

You can remove this again if it's causing you issues -- I just find it convenient to automatically rebuilding the library when something changes in the engine.  To get it to work like this on Windows, you need to copy (or symbolic link) `mingw32-make.exe` to `make.exe`.

happycoder97 replied:
Okay

jpt13653903 committed:
Source/Makefile: Changed "CC" to the more standard "CXX"

Commit:
minor formating

Commit:
Merge branch 'feature/GUI' of github.com:jpt13653903/Gerber2PDF into feature/GUI

jpt13653903 committed:
Refactoring:

- Moved Source to Engine
- Moved the engine rebuild out of the GUI Makefile and into the root Makefile

jpt13653903 commented:
I refactored the build process -- please could you check it?  I don't have all the libraries on my Linux box at work...

Commit:
Merge branch 'feature/GUI' of github.com:jpt13653903/Gerber2PDF into feature/GUI

happycoder97 commented:
Yes it builds

Commit:
Link with dlsym in linux
Handled the edge case when there are no dlls to copy
Fixed a typo in DLLsToCopy

jpt13653903 committed:
Added some ideas to the concept sketch

happycoder97 commented:
This looks excellent :)

happycoder97 commented:
The colour palette shows all used colors, right?

jpt13653903 commented:
> The colour palette shows all used colors, right?

The idea is that there is a standard grid size (make it 6x3, for argument sake) that the user can edit, and the GUI remembers between sessions (see [this card](https://github.com/jpt13653903/Gerber2PDF/projects/1#card-16452909) and [this card](https://github.com/jpt13653903/Gerber2PDF/projects/1#card-16453088)).

One could make it fancy and let the user add more colours, and delete existing ones, but I think that a fixed grid will be good enough for version 1.0...

jpt13653903 committed:
Added notes on how to obtain the local app-data folder path in Windows

jpt13653903 committed:
Updated notes with more logical file-names

jpt13653903 committed:
Changed GUI/Makefile to always re-link the binary

This ensures that the newest version of the Engine is always used, even
if nothing changed in the GUI.

jpt13653903 committed:
Added a clarification to the concept readme

jpt13653903 committed:
Added more detail to the new concept

jpt13653903 committed:
Final tweaks to the new concept sketch

jpt13653903 committed:
Concept update: loading previous configuration

Commit:
Fixed typo in -colour option

Commit:
Merge branch 'feature/GUI' of github.com:jpt13653903/Gerber2PDF into feature/GUI

jpt13653903 committed:
Concept: changed the "mirror" icon

Commit:
Add titlebar and background for gerber list widget

Commit:
Merge branch 'feature/GUI' of github.com:jpt13653903/Gerber2PDF into feature/GUI

happycoder97 commented:
Shouldn't file list have an address bar at the top for manually entering the path?
Does working folder mean the same? or is it where convert.bat and output.pdf are generated?

Commit:
Make strokes2fills is  per file option

Commit:
Update imgui from git

happycoder97 commented:
I'm not sure I handled strokes2fills correctly.. 
Shall I synchronize the checkboxes if a file appears multiple times?
That is, checking one of the boxes will check the boxes of same file appearing elsewhere in the list

Commit:
Increase default window size to 800x600

Commit:
Add basic drag and drop support for gerber list

jpt13653903 commented:
> Shouldn't file list have an address bar at the top for manually entering the path?
> Does working folder mean the same? or is it where convert.bat and output.pdf are generated?

Don't follow the sketch exactly -- that's just me fooling around with ideas ;-)  The finer details are up to you.

My idea is that the working folder is the "current folder", i.e. the folder of the files in the "File List", as well as the location of the output files in the case where they carry no paths.  When the user uses the file list to browse, the working folder updates accordingly.

The user can also save the output files to a different location by specifying a path, or browsing by using the three-dot icons.

It might be a better idea to leave the "working folder" at the top of the "File List", giving the user a stronger association with what it means...

jpt13653903 commented:
> I'm not sure I handled strokes2fills correctly..
> Shall I synchronize the checkboxes if a file appears multiple times?
> That is, checking one of the boxes will check the boxes of same file appearing elsewhere in the list

That is the idea, yes.  All instances of the same file must have the same stroke-to-fills setting.

jpt13653903 commented:
The drag-and-drop is working beautifully :-)  You can drop the "Up" and "Down" button idea.

At some later stage you can make it all nice with shadows and animation and stuff, but for now it's great.

jpt13653903 committed:
Concept: removed the up-down buttons (drag-and-drop is working)

jpt13653903 committed:
Concept: Removed the "browse" button from the working folder

It's pointless to be able to open a browser window when the file list is
the browser window...

Commit:
Synchronize strokes_to_fills option between multiple instances of same file

Commit:
Merge branch 'feature/GUI' of github.com:jpt13653903/Gerber2PDF into feature/GUI

jpt13653903 commented:
> When generating batch script, is it okay if I specify strokes2fills multiple times?
> Like

Yes - specifying it multiple times is fine.

The problem lies in how the Engine does caching.  It reuses the render results of the first instance and simply changes the colour.  If the first instance is filled, all other instances will also be filled, and vice versa.

What I should actually do is change that so that the strokes-to-fills becomes another condition of reuse, in addition to the file name...  Let me see if I can fit that in today...  I'll let you know.

jpt13653903 committed:
Merge branch 'master' into feature/GUI

This merge removes the limitation where -strokes2fills applies the same
setting to all instances of the Gerber file.

jpt13653903 committed:
Concept: Removed the -strokes2fills linking requirement

jpt13653903 commented:
That was much easier than I thought it would be...  You can remove the -strokes2fills linking.  The new version of the Engine applies the `ConvertStrokesToFills` to the next `Run`, and then resets to default immediately afterwards.

Sorry for making you do extra work and then asking you to undo it again half a day later!

happycoder97 commented:
It's okay.. It was only a small work :)

Commit:
Following b68232c, removed strokes2fills linking

Commit:
Merge branch 'feature/GUI' of github.com:jpt13653903/Gerber2PDF into feature/GUI

jpt13653903 committed:
Merge branch 'master' into feature/GUI

Engine update: removed obsolete and duplicate stuff from the PDF output

jpt13653903 committed:
Merge branch 'master' into feature/GUI

Engine updated with new Gerber file format features

jpt13653903 committed:
Merge branch 'master' into feature/GUI

jpt13653903 commented:
Hello,

I'm busy with an update to the Engine and merged in what I've done so far...  I've checked that the Engine compiles under both Windows and Linux, but please also check that it compiles and links correctly on your side.

Nothing changed on the interface you're using.  In summary, I've done the following:

- Pulled in the [Toolbox submodule](https://github.com/jpt13653903/Toolbox)
- Ripped out the obsolete JFile and JString classes in favour of std::string and a better-implemented FileWrapper class (part of the Toolbox submodule)
- Changed the apertures fixed array to std::map (i.e. an associative array)
- Implemented Unicode file-name support (UTF-8 or UTF-16 under Windows and UTF-8 under Linux)

happycoder97 commented:
Hi, I am sorry that I got really busy with college stuff.. 
I hope I will be able to work full time from April onwards, when my vacations begin

jpt13653903 commented:
No worries -- college should take priority ;-)

I'm looking forward to what you come up with during the holidays -- maybe we can launch the first beta-version at the end of it...

jpt13653903 committed:
Merge branch 'master' into feature/GUI

jpt13653903 commented:
Hello,

Long time no see...

Anyway -- I have good news.  T-Dragon has released [MinGW for Windows based on GCC-9](https://jmeubank.github.io/tdm-gcc/articles/2020-03/9.2.0-release).

You should therefore be able to rip out the Boost dependency and use C++17 directly...  When you get a chance, please could you do so?

happycoder97 commented:
Hi,
That sounds awesome!
Will look into it as soon as possible :)

jpt13653903 commented:
I'm closing this as "obsolete"
